File: lib/platform/ios/TodayWidget.swift
import SwiftUI
import WidgetKit

struct TodayEntry: TimelineEntry {
    let date: Date
    let envelope: WidgetEnvelope?
}

/// Reads only the already-computed `today_widget_data` envelope — never
/// calls into any evaluation logic (AC 6, AD-7).
struct TodayProvider: TimelineProvider {
    func placeholder(in context: Context) -> TodayEntry {
        TodayEntry(date: Date(), envelope: nil)
    }

    func getSnapshot(in context: Context, completion: @escaping (TodayEntry) -> Void) {
        completion(TodayEntry(date: Date(), envelope: WidgetDataStore.readEnvelope(key: "today_widget_data")))
    }

    func getTimeline(in context: Context, completion: @escaping (Timeline<TodayEntry>) -> Void) {
        let entry = TodayEntry(date: Date(), envelope: WidgetDataStore.readEnvelope(key: "today_widget_data"))
        // The app reloads this widget's timeline itself (home_widget's
        // WidgetCenter.reloadTimelines call, Story 5.1) whenever it
        // rewrites the shared container — this widget never re-derives its
        // own freshness or polls on a timer.
        completion(Timeline(entries: [entry], policy: .never))
    }
}

/// Today widget (AC 1, 4, 6): reduced-density rendering — each eligible
/// goal's name plus a status dot only, no progress bar, no fraction text
/// (UX-DR18). Story 5.3 (AC1, AC4): `.widgetURL` on the root view sets the
/// widget's single tap destination — `systemSmall` widgets support exactly
/// one tap destination for the whole widget, which matches AC1 exactly
/// (there is only one date to target, so whole-widget granularity is
/// correct here, not a limitation).
struct TodayWidgetEntryView: View {
    var entry: TodayProvider.Entry

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            if entry.envelope == nil || entry.envelope!.isEmpty {
                Text("No goals today").font(.caption)
            } else {
                ForEach(entry.envelope!.cells, id: \.goalId) { cell in
                    HStack(spacing: 8) {
                        StatusCellView(
                            status: cell.status,
                            label: StatusPresentationCatalog.label(goalName: cell.goalName, status: cell.status),
                            size: 16
                        )
                        Text(cell.goalName).font(.caption).lineLimit(1)
                    }
                }
            }
        }
        .padding()
        .widgetURL(entry.envelope.map { WidgetDeepLinks.day($0.generatedAt) })
    }
}

struct TodayWidget: Widget {
    let kind: String = "TodayWidget"

    var body: some WidgetConfiguration {
        StaticConfiguration(kind: kind, provider: TodayProvider()) { entry in
            TodayWidgetEntryView(entry: entry)
        }
        .configurationDisplayName("Today")
        .description("Shows today's goal status at a glance.")
        .supportedFamilies([.systemSmall])
    }
}

File: lib/platform/android/TodayWidget.kt
package com.panda.tracker.tracker.widget

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.glance.GlanceId
import androidx.glance.GlanceModifier
import androidx.glance.action.clickable
import androidx.glance.appwidget.GlanceAppWidget
import androidx.glance.appwidget.SizeMode
import androidx.glance.appwidget.provideContent
import androidx.glance.currentState
import androidx.glance.layout.Alignment
import androidx.glance.layout.Column
import androidx.glance.layout.Row
import androidx.glance.layout.fillMaxWidth
import androidx.glance.layout.padding
import androidx.glance.text.Text
import androidx.glance.text.TextStyle
import com.panda.tracker.tracker.MainActivity
import es.antonborri.home_widget.HomeWidgetGlanceState
import es.antonborri.home_widget.HomeWidgetGlanceStateDefinition
import es.antonborri.home_widget.HomeWidgetGlanceWidgetReceiver
import es.antonborri.home_widget.actionStartActivity

/**
 * Today widget (AC 1, 4, 6): reduced-density rendering — each eligible
 * goal's name plus a status dot only, no progress bars, no fraction text
 * (UX-DR18). Story 5.3 (AC1, AC4): a single click action on the whole
 * widget's root [Column] opens today's Day View — there is only one date to
 * target, so whole-widget granularity is correct here, not a limitation.
 * Reads only the already-computed `status` string from the decoded
 * envelope — no arithmetic, no comparison, no evaluation of any kind (AD-7).
 */
class TodayWidget : GlanceAppWidget() {
    override val sizeMode = SizeMode.Single
    override val stateDefinition = HomeWidgetGlanceStateDefinition()

    override suspend fun provideGlance(context: Context, id: GlanceId) {
        provideContent { Content(context, currentState()) }
    }

    @Composable
    private fun Content(context: Context, state: HomeWidgetGlanceState) {
        val envelope = parseWidgetEnvelope(state.preferences.getString("today_widget_data", null))

        var modifier = GlanceModifier.padding(12.dp)
        if (envelope != null) {
            modifier =
                modifier.clickable(
                    actionStartActivity<MainActivity>(context, WidgetDeepLinks.day(envelope.generatedAt))
                )
        }

        Column(modifier = modifier) {
            if (envelope == null || envelope.isEmpty) {
                Text(text = "No goals today", style = TextStyle(fontSize = 14.sp))
            } else {
                for (cell in envelope.cells) {
                    Row(
                        modifier = GlanceModifier.fillMaxWidth().padding(vertical = 4.dp),
                        verticalAlignment = Alignment.Vertical.CenterVertically,
                    ) {
                        StatusCell(
                            status = cell.status,
                            label = statusLabel(cell.goalName, cell.status),
                            size = 16.dp,
                        )
                        Text(
                            text = cell.goalName,
                            style = TextStyle(fontSize = 14.sp),
                            modifier = GlanceModifier.padding(start = 8.dp),
                        )
                    }
                }
            }
        }
    }
}

class TodayWidgetReceiver : HomeWidgetGlanceWidgetReceiver<TodayWidget>() {
    override val glanceAppWidget = TodayWidget()
}
